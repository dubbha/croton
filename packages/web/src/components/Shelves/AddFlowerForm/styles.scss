@import '../../../custom.scss';

.new-flower {
  &.container{
    justify-content: start;
  }

  h1 {
    color: $color-grass-dark;
  }

  form {
    width: 70%;
    margin-bottom: 76px;
  }

  .btn {
    width: 100%;
  }
}
