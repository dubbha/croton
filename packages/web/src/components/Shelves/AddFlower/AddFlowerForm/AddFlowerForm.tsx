import React, { useState, FormEvent, useEffect } from 'react';
import { Form, SubmitButton, InfoAlert, ErrorAlert, LoadingSpinner } from 'elements';
import { useCustomForm } from 'hooks';
import './styles.scss';

type Props = {
  onSubmit: (name: string, description: string) => void;
  isLoading?: boolean;
  info?: string | null;
  error?: string | null;
};

interface FormFields {
  name?: string;
  description?: string;
}

export const AddFlowerForm = ({
  onSubmit,
  isLoading = false,
  info = null,
  error = null,
}: Props) => {
  const initialValues = {
    name: '',
    description: '',
  };

  const nameErrorMessage = (name: string): string | null => {
    if (name.length < 3 || name.length > 256) {
      return 'Name is required and has to be in range of 3-256 symbols.';
    }
    return null;
  };

  const descriptionErrorMessage = (description: string): string | null => {
    if (description.length < 3 || description.length > 256) {
      return 'Description is required and has to be in range of 3-256 symbols.';
    }
    return null;
  };

  const {
    values,
    errors,
    touched,
    handleChange,
    handleBlur,
  }: {
    values: FormFields;
    errors: FormFields;
    touched: FormFields;
    handleChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
    handleBlur: (event: React.ChangeEvent<HTMLInputElement>) => void;
  } = useCustomForm({
    initialValues,
    errorsRules: {
      name: nameErrorMessage,
      description: descriptionErrorMessage,
    },
  });

  const [isValid, setIsValid] = useState(false);

  useEffect(() => {
    setIsValid(
      !!values.name && !!values.description
      && !errors.name && !errors.description
    );
  }, [errors, values]);

  const handleSubmit = (e: FormEvent) => {
    e.preventDefault();
    if (values.name && values.description) {
      onSubmit(values.name, values.description);
    }
  };

  return (
    <Form onSubmit={handleSubmit} className="flower-form">
      <Form.Group controlId="formName">
        <Form.Label>Flower Name</Form.Label>
        <Form.Control
          type="text"
          placeholder="Enter Flower Name"
          value={values.name}
          name="name"
          onChange={handleChange}
          onBlur={handleBlur}
        />
        {touched.name && errors.name && (
          <Form.Text className="form-error-message">{errors.name}</Form.Text>
        )}
      </Form.Group>

      <Form.Group controlId="formDescription">
        <Form.Label>Description</Form.Label>
        <Form.Control
          type="text"
          placeholder="Enter description"
          value={values.description}
          name="description"
          onChange={handleChange}
          onBlur={handleBlur}
        />
        {touched.description && errors.description && (
          <Form.Text className="form-error-message">{errors.description}</Form.Text>
        )}
      </Form.Group>

      {error && <ErrorAlert>{error}</ErrorAlert>}
      {info && <InfoAlert>{info}</InfoAlert>}
      {!info && !error && (
        <SubmitButton disabled={!isValid}>
          <div className="spinner-container">
            {isLoading && <LoadingSpinner />}
          </div>
          <span>Create</span>
        </SubmitButton>
      )}
    </Form>
  );
};
