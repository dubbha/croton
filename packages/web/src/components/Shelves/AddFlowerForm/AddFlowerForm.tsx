import React, { useState, FormEvent, useEffect } from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faBookmark } from '@fortawesome/free-solid-svg-icons';
import { Form, SubmitButton } from 'elements';
import useCustomForm from 'hooks/useCustomForm';
import './styles.scss';

type Props = {
  onSubmit: (name: string, location: string, description: string) => void;
};

interface FormFields {
  name?: string;
  location?: string;
  description?: string;
}

export const AddFlowerForm = ({ onSubmit }: Props) => {
  const initialValues = {
    name: '',
    location: '',
    description: '',
  };

  const nameErrorMessage = (name: string): string | null => {
    if (name.length < 3 || name.length > 256) {
      return 'Name is required and has to be in range of 3-256 symbols.';
    }
    return null;
  };

  const locationErrorMessage = (location: string): string | null => {
    if (location.length < 3 || location.length > 256) {
      return 'Location is required and has to be in range of 3-256 symbols.';
    }
    return null;
  };

  const {
    values,
    errors,
    touched,
    handleChange,
    handleBlur,
  }: {
    values: FormFields;
    errors: FormFields;
    touched: FormFields;
    handleChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
    handleBlur: (event: React.ChangeEvent<HTMLInputElement>) => void;
  } = useCustomForm({
    initialValues,
    errorsRules: {
      name: nameErrorMessage,
      location: locationErrorMessage,
    },
  });

  const [isValid, setIsValid] = useState(false);

  useEffect(() => {
    setIsValid(
      !!values.name && !!values.location && !!values.description
      && !errors.name && !errors.location && !errors.description
    );
  }, [errors, values]);

  const handleSubmit = (e: FormEvent) => {
    e.preventDefault();
    if (values.name && values.description && values.location) {
      onSubmit(values.name, values.location, values.description);
    }
  };

  return (
    <div className="container new-flower">
      <h1>
        <FontAwesomeIcon icon={faBookmark} />
        <span className="icon-prefix">Add New Flower</span>
      </h1>
      <Form onSubmit={handleSubmit}>
        <Form.Group controlId="formName">
          <Form.Label>Flower Name</Form.Label>
          <Form.Control
            type="text"
            placeholder="Enter Flower Name"
            value={values.name}
            name="name"
            onChange={handleChange}
            onBlur={handleBlur}
          />
          {touched.name && errors.name && (
            <Form.Text className="form-error-message">{errors.name}</Form.Text>
          )}
        </Form.Group>

        <Form.Group controlId="formDescription">
          <Form.Label>Description</Form.Label>
          <Form.Control
            type="text"
            placeholder="Enter description"
            value={values.description}
            name="description"
            onChange={handleChange}
            onBlur={handleBlur}
          />
          {touched.description && errors.description && (
            <Form.Text className="form-error-message">
              {errors.description}
            </Form.Text>
          )}
        </Form.Group>

        <SubmitButton disabled={!isValid}>
          <div className="spinner-container" />
          <span>Create</span>
        </SubmitButton>
      </Form>
    </div>
  );
};
