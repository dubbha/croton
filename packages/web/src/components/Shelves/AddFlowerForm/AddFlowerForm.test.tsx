import React from 'react';
import { render } from '@testing-library/react';
import { AddFlowerForm } from './AddFlowerForm';

describe('components/AddFlowerForm', () => {
  const fn = jest.fn();

  const props = {
    onSubmit: fn
  };

  afterEach(() => {
    fn.mockReset();
  });

  it('should render successfully', () => {
    const { container } = render(<AddFlowerForm {...props} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
