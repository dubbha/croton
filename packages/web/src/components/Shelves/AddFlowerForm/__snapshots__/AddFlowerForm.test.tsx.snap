// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/AddFlowerForm should render successfully 1`] = `
<div
  class="container new-flower"
>
  <h1>
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-bookmark fa-w-12 "
      data-icon="bookmark"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 384 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0 512V48C0 21.49 21.49 0 48 0h288c26.51 0 48 21.49 48 48v464L192 400 0 512z"
        fill="currentColor"
      />
    </svg>
    <span
      class="icon-prefix"
    >
      Add New Flower
    </span>
  </h1>
  <form
    class=""
  >
    <div
      class="form-group"
    >
      <label
        class="form-label"
        for="formName"
      >
        Flower Name
      </label>
      <input
        class="form-control"
        id="formName"
        name="name"
        placeholder="Enter Flower Name"
        type="text"
        value=""
      />
    </div>
    <div
      class="form-group"
    >
      <label
        class="form-label"
        for="formDescription"
      >
        Description
      </label>
      <input
        class="form-control"
        id="formDescription"
        name="description"
        placeholder="Enter description"
        type="text"
        value=""
      />
    </div>
    <button
      class="btn btn-primary"
      data-testid="submitButton"
      disabled=""
      type="submit"
    >
      <div
        class="spinner-container"
      />
      <span>
        Create
      </span>
    </button>
  </form>
</div>
`;
